### PR TITLE
Patterns: Add wrapper to synced patterns and adopt widest alignment of children

### PR DESCRIPTION
**Pull request by aaronrobertshaw**
Fixes:
- https://github.com/WordPress/gutenberg/issues/8288#top

Related:
- https://github.com/WordPress/gutenberg/pull/54170

## What?

- Adds a wrapping element to synced patterns on the front end to match the editor. 
- Adopts full alignment for the pattern, if any of its immediate children have wide/full alignments, to preserve visual appearance

## Why?

It's important to provide a smooth editing experience around patterns. Creating a pattern from wide, full, left, or right-aligned blocks, only to see them move and get squashed into a pattern block wrapper in the editor is unexpected.

## How?

- Adopts layout support on the pattern block (`core/block`)
- Leverages a `pre_render_block` filter to short circuit block rendering when an old pattern doesn't have a wrapper so that layout support styles aren't injected into the first child of the pattern incorrectly
- Now saves a block wrapper element to post content. This is used to distinguish between existing pattern instances that shouldn't have the wrapper added on the frontend and new blocks
- Deprecation added to migrate pattern instances and prevent block invalidation
- The pattern block's edit component infers its alignment from its children and adds the appropriate CSS class to its markup
- The render block callback for the pattern block now temporarily adds a new `render_block_data` filter to determine the widest alignment of its immediate children while they are processed via `do_blocks`
- The `WP_HTML_Tag_Processor` is used to wrap the pattern's entity content with the wrapping element and apply the alignment

## Notes

It was raised in https://github.com/WordPress/gutenberg/pull/54170#discussion_r1316872097 whether we could move the logic to detect the widest alignment to the point where we create the pattern. 

From my explorations, it doesn't look like that or solely relying on updating the pattern's edit component will work. There is a use case where the pattern is created via the block editor and then later edited directly in the `wp_block` post type editor. 

An alternative to the filter approach I've taken here might be to save the alignment to the entity's post meta. I opted against this, as we're already parsing the blocks for the pattern via `do_blocks` in its render callback, so the filters seemed more straightforward for the moment.

#### Backwards Compatibility
1. On trunk, create a synced pattern via the post editor with at least one block with a wide alignment
2. Save the post and inspect on the frontend to confirm no wrapper is applied to the pattern
3. Switch to this PR branch
4. Refresh the frontend page and confirm there is still no wrapper around the pattern content

#### Wrapped Patterns
1. Edit a post, add a series of blocks with varying alignments
5. Select the blocks and via the block toolbar more menu, choose to create a pattern
6. The layout of the blocks should be maintained
7. Save the post and view on the frontend checking that is consistent
8. Repeat the process where the pattern's immediate children have alignment of `wide`, `full`, `left`, or `right`
9. Within a pattern that has any of the alignments from the previous step, remove all such alignments from the pattern's immediate child blocks, then confirm you still have the option to reapply any `wide`, `full`, `left`, or `right` alignment.

## Screenshots or screencast 

https://github.com/WordPress/gutenberg/assets/60436221/2938ff38-16ab-448c-8485-43241331c46e

### CODE REVIEW

**Comment by aaronrobertshaw** [comment]
This isn't quite ready for a proper review yet, I've run out of time but will pick it back up Monday.

**Comment by richtabor** [comment]
Gave it an early look; feels promising :) 

### Before
https://github.com/WordPress/gutenberg/assets/1813435/baee8116-9c63-4ade-8f1e-15a4132a88f7

### After

https://github.com/WordPress/gutenberg/assets/1813435/3511ff5f-b7aa-455b-be2c-7cf4281712d3

**Comment by aaronrobertshaw** [comment]
Most of the layout/alignment quirks have been ironed out I think. So now might be a good time to get a general sanity check on the approach and if there are any obvious issues. 

I'll sort out the failing pattern tests shortly.

File: packages/block-library/src/block/edit.js [hunk 1]
```diff
@@ -45,8 +64,39 @@ export default function ReusableBlockEdit( { attributes: { ref } } ) {
...
+	useEffect( () => {
+		// We only track the initial alignment so that if the user could
+		// initially apply wide or full alignments to inner blocks, that ability
+		// is maintained even if they temporarily remove the last inner block
+		// with that wide/full alignment.
+		if ( ! blocks?.length || inferredAlignment !== undefined ) {
+			return;
+		}
+
+		const isConstrained = __unstableParentLayout.type === 'constrained';
```
Comment (andrewserong) [inline packages/block-library/src/block/edit.js]: `__unstableParentLayout` can (theoretically) be `undefined` ([here](https://github.com/WordPress/gutenberg/blob/fec05ee8451ee1a8da43cdbf00485736e2b0a16a/packages/block-editor/src/components/block-list/block.js#L138)), though in practice it _should_ always fall back to `default` layout type. Just to be safe, shall we add in optional chaining here?

```suggestion
		const isConstrained = __unstableParentLayout?.type === 'constrained';
```

File: packages/block-library/src/block/index.php [hunk 2]
```diff
@@ -41,14 +42,74 @@ function render_block_core_block( $attributes ) {
...
-	return $content;
+
+	remove_filter( 'render_block_data', $filter_alignments, 10 );
+
+	// Older block versions used only the post's content without incorporating
+	// the editor's wrapper. Newer versions added a wrapper through saved
+	// markup, so utilize it to return only the post's content when appropriate.
+	if ( ! $block_content ) {
+		return $content;
+	}
```
Comment (andrewserong) [inline packages/block-library/src/block/index.php:90]: Tiny non-blocking thought: when `$block_content` is empty, we also don't need to add or remove the `render_block_data` filter because `$widest_alignment` won't be used for anything.

**Review by andrewserong — COMMENTED** [review]
This is testing very nicely for me:

✅ Existing unaltered patterns appear to still render correctly without a wrapper introduced on the site frontend, so a pattern inserted into a Row block still renders correctly, for example
✅ Block markup is updated as expected in the editor when making a change
✅ Alignment controls can be adjusted within the editor within the pattern, and setting to None still allows you to set back to Wide / Full

Just left a couple of tiny questions, but overall the direction seems promising to me, and I didn't run into any issues with it!

I see that the tests for the mobile app had to be updated — will this change cause any issues for older versions of the mobile app accessing newer updated markup? 

Edit: as discussed separately this PR still has the issue of layout support classnames being injected into the markup when no wrapper is in use. E.g. for legacy markup in a synced pattern that begins with a heading block, layout classnames get injected into that heading block when they shouldn't be:

image

File: packages/block-library/src/block/edit.js (hunk 1 above)
Comment (aaronrobertshaw) [reply packages/block-library/src/block/edit.js]: Good call, thanks.

File: packages/block-library/src/block/index.php (hunk 2 above)
Comment (aaronrobertshaw) [reply packages/block-library/src/block/index.php:90]: ✅  Done (https://github.com/WordPress/gutenberg/pull/54289/commits/392f6e0ef4027b39ea45bc482fbb4590955ebc17)

**Comment by aaronrobertshaw** [comment]
Thanks for the review, testing, and discussion on this one @andrewserong! It has definitely proven a little tricky to iron out all the edge cases.

> I see that the tests for the mobile app had to be updated — will this change cause any issues for older versions of the mobile app accessing newer updated markup?

That's a good question and one that I need to find the answer to. Unfortunately, I went down a rabbit hole on the other issue you noted and have run out of time today. I'll follow-up on this tomorrow.

> as discussed separately this PR still has the issue of layout support classnames being injected into the markup when no wrapper is in use. E.g. for legacy markup in a synced pattern that begins with a heading block, layout classnames get injected into that heading block when they shouldn't be:

I believe I have something that addresses this issue. In https://github.com/WordPress/gutenberg/pull/54289/commits/d5988733d5c5e1f09638bd69a5e4d3a862fdb232, I've added a `pre_render_block` filter that will short-circuit rendering the pattern block when it doesn't have the saved wrapper. This short-circuiting means the block never gets to the `render_block` filter for the layout supports and as such we don't get the layout styles injected into the first of the pattern's inner blocks.

As the `core/block` type only opts into layout support as of this PR, I don't think we are missing anything. The normal `render_callback` is still in place so the block should also retain its "dynamic" status.

One question/concern I have with this latest approach is whether it is acceptable that any `render_block` filters applied by third parties to this block would now also be skipped for old pattern instances without a wrapper. Perhaps a dev note would suffice?

File: packages/block-library/src/block/index.php [hunk 3]
```diff
@@ -63,3 +128,31 @@ function register_block_core_block() {
...
+
+	$attributes = _wp_array_get( $block, array( 'attrs' ), array() );
+	$inner_html = _wp_array_get( $block, array( 'innerHTML' ), null );
+
+	// Patterns without wrappers will not have saved inner HTML and need to
+	// avoid layout supports applied via render_block.
+	if ( $inner_html ) {
+		return null;
+	}
```
Comment (andrewserong) [inline packages/block-library/src/block/index.php]: Tiny nit: this comment reads like it might better fit just before the `return render_block_core_block` line below. For this `if` block, would a comment like the following work?

```
// For patterns that contain wrappers, skip short-circuiting, and allow block supports to be applied.
```

**Review by andrewserong — COMMENTED** [review]
I like the `pre_render_block` filter approach @aaronrobertshaw, that seems the neatest so far while maintaining behaviour in the two different states of the block, without introducing a new API for this one block's requirements.

It's testing nicely for me, with layout block support no longer being applied to legacy markup 👍

> One question/concern I have with this latest approach is whether it is acceptable that any render_block filters applied by third parties to this block would now also be skipped for old pattern instances without a wrapper. Perhaps a dev note would suffice?

From my perspective, I think of the `core/block` as being kind of a special kind of block — it uses the block API, but isn't really a block that's intended for styling over overriding in a typical way, so short-circuiting for this use case feels appropriate to me. A dev note sounds like a good idea.

That said, I don't have a full understanding of what the potential implications might be, so it'd be well worth getting more folks' perspectives on this one.

To me, though, this seems like a very promising direction and looks good to me!

**Review by tellthemachines — COMMENTED** [review]
Just gave this a quick round of testing before looking through the code in depth. A couple of things:

* Is it intended that any change to a template that has an existing pattern in it will cause a wrapper to go around that pattern? This is a bit unexpected.
* I don't think we want to always apply constrained layout to the wrapper. For instance, if I add a Row to a template (not a post) at root level it's full width by default. Say I add a Site Title and a Navigation to it and set the Row to space between, so now I have a block at each end of the row, which is a pretty common header/footer pattern. If I create a pattern from that Row, its contents jump into the middle of the page. Not only is the Row not full width any longer, I am unable to set it to it full width because the pattern wrapper doesn't support alignments for its children. To make things worse, this is not an editor-only problem: the same happens on the front end. 

I'm sorry I don't have any ready solutions for these problems. My feeling is that it might be better to avoid making changes to the front end for something that is purely an editor view problem, but I'm also very aware of the extensive discussions around this issue and the fact that so far, no good solution has been found 😬

**Comment by tellthemachines** [comment]
I just remembered I had [a thought](https://github.com/WordPress/gutenberg/pull/54170#pullrequestreview-1614499368) when looking at Glen's PR: would the skipping serialization route mentioned in that comment be viable at all? We'd probably still need to set layout dynamically depending on the block context but at least we wouldn't have to change anything on the front end

**Comment by aaronrobertshaw** [comment]
Thanks @andrewserong and @tellthemachines for taking this for a spin.

> Is it intended that any change to a template that has an existing pattern in it will cause a wrapper to go around that pattern? This is a bit unexpected.

I'm open to ideas to avoid it if you have any. 

> I don't think we want to always apply constrained layout to the wrapper. 

Can you provide a video or some example block markup?

>  would the skipping serialization route mentioned in that comment be viable at all?

We would need to conditionally skip serialization and only on the frontend, which would not match other block supports. Could it be done yes but only in just as hacky a way as anything proposed here. Implementing that whole new API for only this special case doesn't seem ideal.

> but at least we wouldn't have to change anything on the front end

The desire is to change things on the frontend though as [suggested](https://github.com/WordPress/gutenberg/pull/54170#discussion_r1317072015) on the original PR. So we are moving towards a consistent display between both front and backends.

**Comment by aaronrobertshaw** [comment]
@youknowriad if you have the time, I'd love to get your thoughts on this PR or whether we should pivot back to pushing https://github.com/WordPress/gutenberg/pull/54170 across the line with some of the lesson's and fixes learnt through exploring this approach.

Trying to force a wrapper for only new pattern instances on the frontend while maintaining layout is proving rather difficult. Short of creating a `core/block-v2` things seem to be getting hackier and hackier, which was why we shelved https://github.com/WordPress/gutenberg/pull/54170 in the first place.

Would you be open to revisiting the earlier approach that does not introduce a wrapper on the frontend, as an interim step? This would provide a smoother pattern creation experience in the editor and buy us time to find a clean solution to introducing the frontend wrapper.

I'm conscious of the fact we need to sort out the pattern editing experience for 6.4 and are running out of time. If there is a better option we can do in a few days, I'm open to new ideas.

**Comment by youknowriad** [comment]
First, I don't think 6.4 pressure should force us to make any quick decision that we might regret later, this issue for reusable blocks have been around for a long time and one release more or less is not going to change a lot. So let's make sure we make the right decision here regardless of any deadline.

1- https://github.com/WordPress/gutenberg/pull/54170 is not that impactful (aside some hacks in the editor), If I'm not wrong it has no impact on the frontend and just "tries" to set the right width of the reusable block wrapper properly. So if we want to merge it, let's make sure that we can revert it without issues/breaking changes... That said, it's very hacky and a bit incomplete. Here are some flows where it fails:
    - I create a reusable block with two blocks (wide and normal), then I change one of the block's width to "full", the reusable block doesn't update its width in this case, it keeps the old "wide" value.
    - What happens when the "layout" definition of the parent of the reusable block changes. Say, we remove the wide width or change the widths. I believe the reusable block will keep the old layout and not reflect that in its children.
    - What happens if the layout type changes (say from "flow" to "flex")
    - Did we address the use-case (less common) where we use "justify-self" on the children of "flex" layouts. There maybe another logic to apply there than just copying the layout from parent to reusable block.

2- Maybe we should create a `core/block-2`? It obviously requires a lot of thinking, but if we think it's the right solution, why not go for it?

**Comment by aaronrobertshaw** [comment]
Thank you for the feedback @youknowriad 🙇 

> First, I don't think 6.4 pressure should force us to make any quick decision that we might regret later

No arguments here!

> https://github.com/WordPress/gutenberg/pull/54170 is not that impactful (aside some hacks in the editor)

That is correct. 

There were a number of use cases it didn't factor in, that were addressed in this PR. I have created a new PR (https://github.com/WordPress/gutenberg/pull/54416) to apply some of these fixes back onto https://github.com/WordPress/gutenberg/pull/54170

Among others, these include:
- Allowing temporary removal of alignments, and not updating the pattern block's layout so alignments can be reapplied
- Adopting the parent layout when the pattern block requires a layout
- Covering left/right alignments etc. Now the pattern block will apply the parent layout if it is constrained and the pattern's inner blocks are any of `full`, `wide`, `left`, or `right`. 

> So if we want to merge it, let's make sure that we can revert it without issues/breaking changes

Using the simple filter to add editor-only layout support and the hook to infer alignment and layout in the block's edit component, is about as low impact as we can make it. Nothing is being persisted with this approach, so reverting it would only bring back the current issue where the pattern contents appear to be constrained when they shouldn't be. 

That visual regression can be mitigated if it's done while applying whatever long-term fix we settle on.

After several long discussions with @glendaviesnz, @tellthemachines, and @andrewserong, my preference would be to proceed with https://github.com/WordPress/gutenberg/pull/54170 once the fixes from https://github.com/WordPress/gutenberg/pull/54416 and any others required have been applied. 

That improves the editing experience around patterns and buys us time to think through the best long term option, including whether that is a v2 of the `core/block`.

> That said, it's very hacky and a bit incomplete. Here are some flows where it fails

I'll give https://github.com/WordPress/gutenberg/pull/54416 some further testing tomorrow, working through the flows identified.

Very much appreciate the advice shared, thank you 👍

**Comment by youknowriad** [comment]
Ok sounds good to me. Thanks for considering the alternative, I appreciate it. It does make sense to try the non-impactful approach first. I left a small comment here https://github.com/WordPress/gutenberg/pull/54416#discussion_r1324330905 that could improve things too.

**Comment by aaronrobertshaw** [comment]
Closing this PR in favour of https://github.com/WordPress/gutenberg/pull/54416 and https://github.com/WordPress/gutenberg/pull/54170
